### PR TITLE
apply isort, black, and flake8 utilities on Python scripts

### PR DESCRIPTION
**Pull request by tristan0x**
Also add `dev/python.sh` executable to check or format Python code

### CODE REVIEW

**Comment by WeiliangChenOIST** [comment]
@tristan0x looks like they are all format changes, correct?

**Comment by tristan0x** [comment]
Flake8 changes are a little bit subtle than that but at the end, they do not change execution. To give you an example, Flake8 raises an error whenever PEP8 E713 rule is broken. This rule states that one should not write `if not e in l` but `if e not in l` instead. And they are dozens of rules like this.

For black and isort, it is only formatting changes.

**Review by WeiliangChenOIST — APPROVED** [review]

**Comment by iahepburn** [comment]
Just being honest, personally I feel a little bit uncomfortable with these kind of automated changes. For example, in the tutorials things are put in a specific order because that's the way they are introduced e.g. a module might be imported halfway through because it's introduced at that point with discussion about how it will be used there, but now all modules are imported at the top of the script. That might be neater in a script but affects the flow of a tutorial. And otherwise it looks like just whitespace changes but can we be 100% confident in this, no chance whatsoever of some error being introduced? Some of the publication scripts are exactly that - published- and to change them now feels kind of bad practice to me even if they are just copies of things published on modeldb. I'd rather keep published scripts as they are. These are just my honest opinions.

**Comment by WeiliangChenOIST** [comment]
I think we can apply this formatting rule on the `examples` folder (currently `python_scripts` and `notebooks`), and keep the other folder content as it is. @iahepburn please check if there is any script in `python_scripts` cannot be reformatted. @tristan0x is there anyway to reverse the changes on folders except `python_scripts`?

**Comment by iahepburn** [comment]
Another idea- and probably my personal preference- is to do something similar to SBML biomodels etc to separate scripts into 'curated' and 'non-curated' or something like that, and then only put scripts into 'curated' that are verified to work on a particular version of STEPS and any recent additions should match the new style too. So, for example, the 2013 calcium burst scripts, without changes, could go into 'curated/STEPS2.0' (or whatever the version was- we can check in the paper). If someone wants to put them also into 'curated/STEPS3.4' then we can apply the new style but then someone would also need to check output. Right now I am manually checking what's presently in 'python_scripts' under the new style and if that goes well then I'd be happy for them to go into 'curated/STEPS3.4'.

**Comment by iahepburn** [comment]
So I have got this far with verifying things in python_scripts on STEPS 3.4:

- [x] diffusion
- [x] diffusion_boundary
- [x] directional_dcst
- [x] HH_APprop
- [x] ip3
- [x] sbml
- [x] surface_diffusion
- [x] surface_diffusion_boundary
- [x] well_mixed
- [x] parallel
- [x] visual

So I couldn't get visual working- @WeiliangChenOIST is going to look into that because it's his baby.
 
The problem I get with the parallel script is this, let me just copy it here:
Fatal error in MPI_Init: Other MPI error, error stack:
MPIR_Init_thread(565)..............: 
MPID_Init(224).....................: channel initialization failed
MPIDI_CH3_Init(105)................: 
MPID_nem_init(324).................: 
MPID_nem_tcp_init(178).............: 
MPID_nem_tcp_get_business_card(425): 
MPID_nem_tcp_init(384).............: gethostbyname failed, Iains-MacBook-Pro.local (errno 1)

Basically I can't test the parallel script at all, so I'd like someone else to try. May just be some issue with my mpi installation, but I'm not going to be able to figure that out before end of play today. 
Otherwise, all the checked things, I'm happy with.

**Comment by WeiliangChenOIST** [comment]
having mpi+petsc+openblas issues too...

**Comment by tristan0x** [comment]
> @tristan0x is there anyway to reverse the changes on folders except `python_scripts`?

It is done.

**Comment by tristan0x** [comment]
> having mpi+petsc+openblas issues too...

Despite these issues, this PR can be merged right?

**Comment by iahepburn** [comment]
> > having mpi+petsc+openblas issues too...
> 
> Despite these issues, this PR can be merged right?

Not for me. I feel my comments haven't been answered yet, and we are both having mpi issues this end so that we can't verify the scripts.

**Comment by tristan0x** [comment]
> but now all modules are imported at the top of the script. That might be neater in a script but affects the flow of a tutorial.

To me, symbol imports and usage are uncorrelated. On the contrary, it makes more sense to me to know right at the top of the script what will be used.

> And otherwise it looks like just whitespace changes but can we be 100% confident in this, no chance whatsoever of some error being introduced?

isort and black are used by thousands of people, and they sanitize billions of LOC every day. I am pretty confident about them.

> I'd rather keep published scripts as they are.

This PR does not modify these scripts.

**Comment by tristan0x** [comment]
> > > having mpi+petsc+openblas issues too...
> > 
> > 
> > Despite these issues, this PR can be merged right?
> 
> Not for me. I feel my comments haven't been answered yet, and we are both having mpi issues this end so that we can't verify the scripts.

This is an OpenMP issues in STEPS (see CNS-OIST/HBP_STEPS#169) so not related to this PR, right @WeiliangChenOIST ?

**Comment by iahepburn** [comment]
> > but now all modules are imported at the top of the script. That might be neater in a script but affects the flow of a tutorial.
> 
> To me, symbol imports and usage are uncorrelated. On the contrary, it makes more sense to me to know right at the top of the script what will be used.
> 
Ok- it's quite a minor issue so I'm not going to push too hard on this, but just be aware that these scripts now don't match the tutorial perfectly historically nor currently now in the jupyter notebooks. 


> > And otherwise it looks like just whitespace changes but can we be 100% confident in this, no chance whatsoever of some error being introduced?
> 
> isort and black are used by thousands of people, and they sanitize billions of LOC every day. I am pretty confident about them.
> 
> > I'd rather keep published scripts as they are.
> 
> This PR does not modify these scripts.

Thanks for reversing this. I definitely feel we shouldn't modify published scripts even if they are not very beautiful. If we do modify them - which I am not against per say- then we need to check them in the current version which would be months of work for the J Neurosci paper, though less work for others. 


Still would like to verify the visual and parallel scripts before merging if possible- though not if it's weeks of work obviously. If we can't get the visual one working at all, for example, then I think we should remove it for now. 

What do you both think about separating the scripts into STEPS versions? We can't be sure, for example, the J Neurosci ones would work on STEPS 3.4 without modification anyway.

**Comment by WeiliangChenOIST** [comment]
visual and parallel are tested with Ubuntu (still have mpi issue with Mac OS).

I think the main discussion now is on the layout of this repository i.e. whether we associated the scripts with particular STEPS version.

In my opinion it is necessary to provide the running version for each script, but to reduce reduncency we only need to keep the latest supported version. For instance, if the J Neurosci scripts was tested and passed for both 2.0 and 3.0, we will keep 3.0 only. But if some other script fails at 3.0 with various reasons, then we will keep the 2.0 version.

Since now this PR only affects the scripts in python_scripts, and effectly bumps their supported STEPS version to HBP_STEPS master, we can probably merge this PR first, then have another PR specifically for the overall layout changes including version curation. 

@iahepburn opinion?

**Comment by WeiliangChenOIST** [comment]
> > > > having mpi+petsc+openblas issues too...
> > > 
> > > 
> > > Despite these issues, this PR can be merged right?
> > 
> > 
> > Not for me. I feel my comments haven't been answered yet, and we are both having mpi issues this end so that we can't verify the scripts.
> 
> This is an OpenMP issues in STEPS (see [CNS-OIST/HBP_STEPS#169](https://github.com/CNS-OIST/HBP_STEPS/issues/169)) so not related to this PR, right @WeiliangChenOIST ?

It is not related.

**Comment by iahepburn** [comment]
> visual and parallel are tested with Ubuntu (still have mpi issue with Mac OS).
> 
> I think the main discussion now is on the layout of this repository i.e. whether we associated the scripts with particular STEPS version.
> 
> In my opinion it is necessary to provide the running version for each script, but to reduce reduncency we only need to keep the latest supported version. For instance, if the J Neurosci scripts was tested and passed for both 2.0 and 3.0, we will keep 3.0 only. But if some other script fails at 3.0 with various reasons, then we will keep the 2.0 version.
> 
> Since now this PR only affects the scripts in python_scripts, and effectly bumps their supported STEPS version to HBP_STEPS master, we can probably merge this PR first, then have another PR specifically for the overall layout changes including version curation.
> 
> @iahepburn opinion?

I'm onboard with all of this, though let's not forget to do the layout change in a later PR then if we don't do it here. Yes- let's only keep the most recent verified script, though I have to say I think it's quite unlikely we'll ever verify the J Neurosci stuff for STEPS 3 because it would be months of work. 

In general, I am fine with enforcing more standardised code from now on both in terms of the Python scripts and C++ code. I am on board with this recent initiative from BBP, but our work is 2-pronged and from a scientific perspective it's also important to verify that the scripts run and are producing the results that we say they do. Remember I am basically Erik's spokesman on this - I know that's what he will say. That's why we need to check any changes we make to model scripts, and - as were mentioned originally and now I am more on board with- I think we should come up with a set of guidelines for verifying/curating model scripts. This is particularly important for anything published - we need to check that published results are reproducible with these scripts for any changes either in the Python script itself and/or new STEPS versions that may have new solvers or methods. 

I still can't get the visual script to work and as a side note to all this I feel it may need a bit of maintenance. Will discuss with @WeiliangChenOIST .

**Comment by WeiliangChenOIST** [comment]
@tristan0x I will run some more tests with visual on Mac now, after that we can merge this PR.

**Comment by WeiliangChenOIST** [comment]
visual works fine with anaconda + Mac. Also parallel seems working fine with anaconda again. running further tests.
Parallel test passed with anaconda + Mac. @tristan0x I am fine for the merging. @iahepburn after that I will initial a PR branch with layout changes, will discuss with you then.
